src/internal/RangeCounting.cpp

```diff
@@ -0,0 +1,286 @@
+#include <tvm/internal/RangeCounting.h>
+
+namespace tvm::internal
+{
+bool RangeCounting::add(const Range & r)
+{
+  assert(isValid());
+  if(r.dim == 0)
+    return false;
+
+  recomputeSplit_ = true;
+  if(limits_.empty())
+  {
+    limits_.emplace_back(r.start, Limit::Lower);
+    limits_.emplace_back(r.end(), Limit::Upper);
+    return recompute(true);
+  }
+  else
+  {
+    int depth = 0;
+    bool change = false;
+    Limit l = {r.start, Limit::Lower};
+    Limit u = {r.end(), Limit::Upper};
+    auto it = limits_.begin();
+    moveToFirstAfter(l, it, depth);
+    if(it == limits_.end())
+    {
+      limits_.push_back(l);
+      limits_.push_back(u);
+      return recompute(true);
+    }
+    insert(l, it, depth);
+    if(depth == 1)
+      change = true;
+
+    change = moveToFirstAfter(Limit(r.end(), Limit::Cut), it, depth, 1) || change;
+    insert(u, it, depth);
+    return recompute(change);
+  }
+}
+
+bool RangeCounting::remove(const Range & r)
+{
+  assert(isValid());
+  if(r.dim == 0)
+    return false;
+
+  if(limits_.empty())
+    throw std::runtime_error("Cannot remove interval on empty list");
+  if(r.start < limits_.front().i_ || r.start >= limits_.back().i_ || r.end() <= limits_.front().i_
+     || r.end() > limits_.back().i_)
+    throw std::runtime_error("The removed interval must be included in the current intervals");
+
+  recomputeSplit_ = true;
+  int depth = 0;
+  Limit l(r.start, Limit::Upper);
+  Limit u(r.end(), Limit::Lower);
+  auto it = limits_.begin();
+  moveToFirstAfter({r.start, Limit::Cut}, it, depth);
+  if(depth == 0)
+    throw std::runtime_error("Trying to remove from a non-existing interval");
+  bool change = (depth == 1);
+  auto it1 = it; // We don't insert the value yet, to keep a valid state in case a exception is thrown
+  int depth1 = depth;
+  change = moveToFirstAfter(u, it, depth, 1) || change;
+  insert(l, it1, depth1);
+  --depth;
+  insert(u, it, depth);
+  depth = 0;
+  it = limits_.begin();
+  while(it != limits_.end()) // ideally from it1 to it, but the second insert can invalidate it1
+  {
+    depth -= it->type_;
+    if(depth == 0 && it->type_ == Limit::Cut)
+      it = limits_.erase(it);
+    else
+      ++it;
+  }
```

Comment:
This could use some comments, I don't understand how this works

Reply:
Hopefully, the additional comments in 32599c3 should do it.